File: src/Grid.ts
import { Colour } from "./types.js";

class Grid<T> {

    public width: number;
    public height: number;
    public fillValue: T;
    public grid: T[][];

    constructor(width: number, height: number, fillValue: T) {

        this.width = width
        this.height = height
        this.fillValue = fillValue;
        this.grid = new Array(width).fill([]).map(e => new Array(height).fill(fillValue));

    }

    public static createGrid<T>(width: number, height: number, transformer: (x: number, y: number) => T, fill: T): Grid<T> {
        let output = new Grid<T>(width, height, fill);
        for (let x = 0; x < width; x++) {
            for (let y = 0; y < height; y++) {
                output.setValueAt(x, y, transformer(x, y));
            }
        }
        return output;
    }

    public valueAt(x: number, y: number): T {
        return this.grid[x][y];
    }

    public setValueAt(x: number, y: number, value: T): void {
        this.grid[x][y] = value;
    }

    public isInBounds(x: number, y: number): boolean {
        return (0 <= x && x < this.width && 0 <= y && y < this.height);
    }

    public overlay(overlayGrid: Grid<T>, cornerX: number = 0, cornerY: number = 0): void {
        for (let _x = 0; _x < overlayGrid.width; _x++) {
            for (let _y = 0; _y < overlayGrid.height; _y++) {
                let x = cornerX + _x;
                let y = cornerY + _y;
                let value = overlayGrid.valueAt(_x, _y);
                if (0 <= x && x < this.width && 0 <= y && y < this.height) {
                    this.setValueAt(x, y, value);
                }
            }
        }
    }

    public static getHeightColour(height: number): Colour {
        let v = Math.floor((1 - height) * 255);
        return [v, v, v];
    }

    public copy(): Grid<T> {
        let grid = new Grid<T>(this.width, this.height, this.fillValue);
        for (let x = 0; x < this.width; x++) {
            for (let y = 0; y < this.height; y++) {
                let v = this.valueAt(x, y);
                grid.setValueAt(x, y, v);
            }
        }
        return grid;
    }

    public static calculateAverage(grid: Grid<number>): number {
        let acc = 0
        for (let x = 0; x < grid.width; x++) {
            for (let y = 0; y < grid.height; y++) {
                acc += grid.valueAt(x, y) as number;
            }
        }
        return acc / (grid.width * grid.height);
    }

    public printAllValues(): void {
        for (let y = 0; y < this.height; y++) {
            for (let x = 0; x < this.width; x++) {
                console.log(this.valueAt(x, y));
            }
        }
    }

}

export default Grid;

File: src/Grid.py
import os
from PIL import Image

class Grid:

    def __init__(self, width, height, fill):

        self.width = width
        self.height = height
        self.grid = [[fill] * height for _ in range(width)]

    def value_at(self, x, y):
        return self.grid[x][y]

    def set_value_at(self, x, y, value):
        self.grid[x][y] = value

    def is_in_bounds(self, x, y):
        return (0 <= x < self.width and 0 <= y < self.height)

    def overlay(self, overlay_grid, corner_x = 0, corner_y = 0):
        for _x in range(overlay_grid.width):
            for _y in range(overlay_grid.height):
                x, y = corner_x + _x, corner_y + _y
                value = overlay_grid.value_at(_x, _y)
                if 0 <= x < self.width and 0 <= y < self.height:
                    self.set_value_at(x, y, value)

    @staticmethod
    def get_height_colour(height):
        v = int((1 - height) * 255)
        return (v, v, v)

    def save_image(self, filename, WORLD_NAME):
        img = Image.new("RGB", (self.width, self.height), "black")

        pixels = img.load()

        for x in range(self.width):
            for y in range(self.height):
                v = self.grid[x][y]
                rgb = Grid.get_height_colour(v)
                pixels[x, y] = rgb

        img.save(os.path.join("worlds", WORLD_NAME, "GENERATED", "images", filename + ".png"))

    def save_RGBs(self, filename, WORLD_NAME):
        img = Image.new("RGB", (self.width, self.height), "black")

        pixels = img.load()

        for x in range(self.width):
            for y in range(self.height):
                pixels[x, y] = self.grid[x][y]

        img.save(os.path.join("worlds", WORLD_NAME, "GENERATED", "images", filename + ".png"))

    def copy(self):
        grid = Grid(self.width, self.height, 0)

        for x in range(self.width):
            for y in range(self.height):
                v = self.value_at(x, y)
                grid.set_value_at(x, y, v)

        return grid

    def calculate_average(self):
        acc = 0
        for x in range(self.width):
            for y in range(self.height):
                acc += self.value_at(x, y)

        return acc / (self.width * self.height)

